### PR TITLE
Transfer proxy env vars from hive-operator pod to all hive/provision …

### PR DESCRIPTION
**Pull request by dgoodwin**
…pods.

These env vars should be set by OLM if a cluster is configured with a
global proxy. This change ensures they are passed through to all pods
the operator creates, and all pods hive-controllers creates.

x-ref: https://issues.redhat.com/browse/HIVE-1355

/assign @abhinavdahiya 
/cc @abutcher

### CODE REVIEW

**Comment by dgoodwin** [comment]
Missed ImageSet jobs, adding now.

**Comment by dgoodwin** [comment]
Added some tests and refactored how and where we read env vars. Ready for review @abhinavdahiya

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 1]
```diff
@@ -278,6 +281,8 @@ type ReconcileClusterDeployment struct {
 	validateCredentialsForClusterDeployment func(client.Client, *hivev1.ClusterDeployment, log.FieldLogger) (bool, error)
 
 	protectedDelete bool
+
+	httpProxy, httpsProxy, noProxy string
```
Comment (abhinavdahiya) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:285]: not a big fan of storing the proxy information in the reconciler. Go binaries automatically support reading these env vars and using them for http client requests, so all reconciler calls will be covered.
and only thing we need to do it copy these env vars to other pods we create.

So i would like to see that, us copying env vars to pods we created and not using reconciler fields.

File: pkg/controller/utils/utils.go [hunk 2]
```diff
@@ -350,3 +350,38 @@ func BuildControllerLogger(controller hivev1.ControllerName, resource string, ns
 		"reconcileID": utilrand.String(constants.ReconcileIDLen),
 	})
 }
+
+// SetProxyEnvVars will add the standard proxy environment variables to all containers in the given pod spec. If
+// any of the provided values are empty, the environment variable will not be set.
+func SetProxyEnvVars(podSpec *corev1.PodSpec, httpProxy, httpsProxy, noProxy string) {
+	setEnvVarOnContainers := func(podSpec *corev1.PodSpec, envVar, value string) {
+		if value != "" {
```
Comment (abhinavdahiya) [inline pkg/controller/utils/utils.go:358]: return early

File: pkg/controller/utils/utils.go [hunk 3]
```diff
@@ -350,3 +350,38 @@ func BuildControllerLogger(controller hivev1.ControllerName, resource string, ns
...
+			for i := range podSpec.Containers {
+				// Check if the env var is already set, if so we preserve the original but warn if they differ:
```
Comment (abhinavdahiya) [inline pkg/controller/utils/utils.go:360]: this sounds off from the goal of the SetProxyEnvVars .. i would assume we force our value.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 1 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:285]: This was for testability reasons and we'd talked in the past about avoiding reading env vars directly in controllers, but I could live with the controllers doing Getenv, skip testing at that level, and pass the strings into SetProxyEnvVars which can still be unit tested cleanly. I'll make the change.

**Comment by dgoodwin** [comment]
Updates pushed.

File: pkg/operator/hive/hive_controller.go [hunk 4]
```diff
@@ -69,7 +69,15 @@ func Add(mgr manager.Manager) error {
...
-	return &ReconcileHiveConfig{Client: mgr.GetClient(), scheme: mgr.GetScheme(), restConfig: mgr.GetConfig(), mgr: mgr}
+	return &ReconcileHiveConfig{
+		Client:     mgr.GetClient(),
+		scheme:     mgr.GetScheme(),
+		restConfig: mgr.GetConfig(),
+		mgr:        mgr,
+		httpProxy:  os.Getenv("HTTP_PROXY"),
+		httpsProxy: os.Getenv("HTTPS_PROXY"),
+		noProxy:    os.Getenv("NO_PROXY"),
+	}
```
Comment (abhinavdahiya) [inline pkg/operator/hive/hive_controller.go:80]: i think this missed the reconciler to callsite change

Comment (dgoodwin) [reply pkg/operator/hive/hive_controller.go]: My bad! Forgot that area entirely.

**Pull request merged**